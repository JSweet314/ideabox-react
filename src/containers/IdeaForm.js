import React, {Component} from 'react';
import Idea from '../Idea.js';
import '../IdeaForm.css';
import PropTypes from 'prop-types';

class IdeaForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      titleInput: '',
      bodyInput: '',
      saveDisabled: true
    };
  }

  updateTitleInput(e) {
    const waitingForUserInput = e.target.value && this.state.bodyInput;

    this.setState({
      titleInput: e.target.value,
      saveDisabled: !waitingForUserInput
    });
  }

  updateBodyInput(e) {
    const waitingForUserInput = e.target.value && this.state.titleInput;

    this.setState({
      bodyInput: e.target.value,
      saveDisabled: !waitingForUserInput
    });
  }

  createIdea(event) {
    event.preventDefault();
    const idea = new Idea(this.state.titleInput, this.state.bodyInput);

    this.props.handleNewIdea(idea);
    this.setState({
      titleInput: '',
      bodyInput: '',
      saveDisabled: true
    });
  }

  render() {
    return (
      <form>
        <h1><span>idea</span>box</h1>
        <input 
          aria-label="idea title input"
          className="titleInput"
          type="text"
          placeholder="Idea Title"
          value={this.state.titleInput}
          onChange={(e) => this.updateTitleInput(e)}
        />
        <input 
          aria-label="idea body input"
          className="bodyInput"
          type="text"
          placeholder="Idea Body"
          value={this.state.bodyInput}
          onChange={(e) => this.updateBodyInput(e)}
        />
        <button 
          title="Save Idea"
          onClick={(e) => this.createIdea(e)}
          disabled={this.state.saveDisabled}>
          save
        </button>
      </form>
    );
  }
}

IdeaForm.propTypes = {
  handleNewIdea: PropTypes.func.isRequired
};

export default IdeaForm;