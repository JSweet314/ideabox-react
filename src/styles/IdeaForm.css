form {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  background-color: #e5f3f2;
  color: #6d6e71;
  font-family: 'Open Sans', sans-serif;
  width: 100%;
  text-align: center;
}

h1 {
  color: #6d6e71;
  font-family: 'Roboto Slab', serif;
  font-size: 2rem;
}

h1 span {
  color: #00a79d;
}

form input {
  line-height: 1;
  padding: 0 0 0 0.5rem;
}

form input,
form button {
  width: 50%;
  margin: 0 0 1rem 0;
  height: 2rem;
  border: 2px solid lightgray;
  font-size: 1.2rem;
  font-family: 'Open Sans', sans-serif;
}

form button {
  background-color: #00a79d;
  color: white;
  cursor: pointer;
  width: 50%;
  border: none;
}

form button:hover {
  background-color: #045e55;
}