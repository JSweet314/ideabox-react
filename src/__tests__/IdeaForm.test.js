import React from 'react';
import { shallow, mount } from 'enzyme';
import IdeaForm from '../containers/IdeaForm.js';

describe('IdeaForm', () => {
  let ideaForm, event1, event2, event3;
  
  beforeEach(() => {
    event1 = {
      target: {
        value: "Woo Hoo!"
      }
    };

    event2 = {
      target: {
        value: "Hello World!"
      }
    };

    event3 = {
      preventDefault: jest.fn()
    };

    ideaForm = shallow(<IdeaForm handleNewIdea={jest.fn()} />);
  });
  
  it('should render a form to obtain user input', () => {
    expect(ideaForm).toMatchSnapshot();
  });
  
  it('should store the values of the title input and body input and whether the save button is disabled in its state', () => {
    expect(ideaForm.state()).toEqual({
      titleInput: '',
      bodyInput: '',
      saveDisabled: true
    });
  });

  it('should be able to update the title and body in state', () => {
    expect(ideaForm.state('titleInput')).toEqual('');
    expect(ideaForm.state('bodyInput')).toEqual('');

    ideaForm.instance().updateTitleInput(event1);
    ideaForm.instance().updateBodyInput(event2);

    expect(ideaForm.state('titleInput')).toEqual('Woo Hoo!');
    expect(ideaForm.state('bodyInput')).toEqual('Hello World!');
  });

  it('should update title and body as the user types', () => {
    expect(ideaForm.state('titleInput')).toEqual('');
    expect(ideaForm.state('bodyInput')).toEqual('');

    const titleInput = ideaForm.find('.titleInput');
    const bodyInput = ideaForm.find('.bodyInput');

    titleInput.simulate('change', event1);
    bodyInput.simulate('change', event2);

    expect(ideaForm.state('titleInput')).toEqual('Woo Hoo!');
    expect(ideaForm.state('bodyInput')).toEqual('Hello World!');
  });

  it('should enable to save button when both the title and body input fields are completed', () => {
    expect(ideaForm.state('saveDisabled')).toEqual(true);
    
    ideaForm.instance().updateTitleInput(event1);

    expect(ideaForm.state('saveDisabled')).toEqual(true);

    ideaForm.instance().updateBodyInput(event2);

    expect(ideaForm.state('saveDisabled')).toEqual(false);

    event1.target.value = '';
    ideaForm.instance().updateTitleInput(event1);

    expect(ideaForm.state('saveDisabled')).toEqual(true);
  });

  it('should call its create idea method when the save button is clicked', () => {
    ideaForm = mount(<IdeaForm handleNewIdea={jest.fn()} />);
    ideaForm.instance().updateTitleInput(event1);
    ideaForm.instance().updateBodyInput(event2);
    ideaForm.instance().createIdea = jest.fn();

    const saveBtn = ideaForm.find('form button');

    saveBtn.simulate('click');

    expect(ideaForm.instance().createIdea).toHaveBeenCalledTimes(1);
  });

  it('should call the handleNewIdea method when the save button is clicked', () => {
    ideaForm = mount(<IdeaForm handleNewIdea={jest.fn()} />);
    ideaForm.instance().updateTitleInput(event1);
    ideaForm.instance().updateBodyInput(event2);
    const saveBtn = ideaForm.find('form button');

    saveBtn.simulate('click', event3);
    
    expect(ideaForm.props().handleNewIdea).toHaveBeenCalledTimes(1);
    expect(ideaForm.state()).toEqual({
      titleInput: '',
      bodyInput: '',
      saveDisabled: true
    });
  });
});
