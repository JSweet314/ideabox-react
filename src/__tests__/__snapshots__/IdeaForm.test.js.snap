// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdeaForm should render a form to obtain user input 1`] = `
<form>
  <h1>
    <span>
      idea
    </span>
    box
  </h1>
  <input
    aria-label="idea title input"
    className="titleInput"
    onChange={[Function]}
    placeholder="Idea Title"
    type="text"
    value=""
  />
  <input
    aria-label="idea body input"
    className="bodyInput"
    onChange={[Function]}
    placeholder="Idea Body"
    type="text"
    value=""
  />
  <button
    disabled={true}
    onClick={[Function]}
    title="Save Idea"
  >
    save
  </button>
</form>
`;
